tools/cwtool/TxSimulator.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module TxSimulator
+  where
+
+import Control.Monad.Catch
+import Data.Maybe
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import qualified Data.Vector as V
+import Options.Applicative
+import System.LogLevel
+
+import Chainweb.BlockHeader
+import Chainweb.Logger
+import Chainweb.Pact.Backend.RelationalCheckpointer
+import Chainweb.Pact.Backend.Types
+import Chainweb.Pact.Backend.Utils
+import Chainweb.Pact.PactService
+import Chainweb.Pact.RestAPI.Server
+import Chainweb.Pact.TransactionExec
+import Chainweb.Pact.Types
+import Chainweb.Payload
+import Chainweb.Time
+import Chainweb.Transaction
+import Chainweb.Utils
+import Chainweb.Version
+
+import Pact.Types.Command
+import Pact.Types.Logger
+import Pact.Types.SPV
+
+data SimConfig = SimConfig
+    { scDbDir :: FilePath
+      -- ^ db dir containing sqlite pact db files
+    , scParentBlock :: FilePath
+      -- ^ block header file.
+    , scPayload :: FilePath
+      -- ^ payload file
+    , scTxIndex :: Int
+      -- ^ index in payload transactions list
+    , scChain :: ChainId
+    , scVersion :: ChainwebVersion
+    }
+
+simulate :: SimConfig -> IO ()
+simulate (SimConfig dbDir parentBlockFile payloadFile txIdx cid ver) = do
+  parent <- decodeFileStrictOrThrow parentBlockFile
+  ((PayloadData txs md _ _ _) :: PayloadData) <- decodeFileStrictOrThrow payloadFile
+  let (Transaction tx) = txs V.! txIdx
+  cmdTx <- decodeStrictOrThrow tx
+  miner <- decodeStrictOrThrow $ _minerData md
+  case validateCommand cmdTx of
+    Left _ -> error "bad cmd"
+    Right cmdPwt ->
+      withSqliteDb cid cwLogger dbDir False $ \sqlenv -> do
+        (CheckpointEnv cp _) <- initRelationalCheckpointer (initBlockState 0) sqlenv logger ver cid
+        bracket
+          (_cpBeginCheckpointerBatch cp)
+          (\_ -> _cpDiscardCheckpointerBatch cp) $ \_ -> do
+            let cmd = payloadObj <$> cmdPwt
+                txc = txContext parent cmd
+            (PactDbEnv' pde) <- _cpRestore cp $ Just (succ (_blockHeight parent), _blockHash parent)
```

Comment:
Ok, going full nerd now: Why the parens on the left side?

Reply:
I de-parenthesized all such expressions :)